File: programmers/algorithm_category/sung-silver/dfs-bfs/week8-1.java
class Solution {
    int answer=0;
    public int solution(int[] numbers, int target) {
        answer = 0;
        dfs(0,0,numbers,target);
        return answer;
    }
    
    void dfs(int sum, int location, int[] numbers, int target){
        if(location == numbers.length){
            if(sum == target){
                answer++;
            }
            return ;
        }
        dfs(sum+numbers[location], location+1, numbers, target);
        dfs(sum-numbers[location], location+1, numbers, target);
    }
}

File: programmers/algorithm_category/Hong0329/dfs-bfs/week8-1.cpp
//
// https://school.programmers.co.kr/learn/courses/30/lessons/43165
//
#include <string>
#include <vector>

using namespace std;

int answer = 0;

void dfs(vector<int> numbers, int target, int sum, int num) {
    if(num == numbers.size()) {
        if(sum == target) {
            answer++;
        }
        return;
    }
    dfs(numbers, target, sum + numbers[num], num + 1);
    dfs(numbers, target, sum - numbers[num], num + 1);
}

int solution(vector<int> numbers, int target) {
    dfs(numbers, target, 0, 0);
    return answer;
}